File: src/lib/api/exposure.ts
import type { ExposureProjection, RedactionProof } from "../types";
import { call, isTauriRuntime } from "./shared";

const demoExposure: ExposureProjection = {
  state: "scanned",
  deep: false,
  filesScanned: 128,
  durationMs: 41,
  counts: { certain: 3, likely: 1, possible: 0, allowed: 1, managed: 1 },
  findings: [
    {
      path: ".env.local",
      kind: "env-file",
      severity: "certain",
      disposition: "managed",
    },
    {
      path: ".env.development",
      kind: "env-file",
      severity: "certain",
      disposition: "allowed",
      reason: "ai-allowed-variable",
    },
    {
      path: "credentials.json",
      kind: "credential-file",
      severity: "certain",
      disposition: "exposed",
    },
    {
      path: "apps/web/.npmrc",
      kind: "npmrc",
      severity: "certain",
      disposition: "exposed",
    },
    {
      path: ".cursor/mcp.json",
      kind: "mcp-config",
      severity: "likely",
      disposition: "exposed",
    },
  ],
};

export async function scanExposure(
  projectId: string,
  deep = false,
): Promise<ExposureProjection> {
  if (!isTauriRuntime) {
    return { ...demoExposure, deep };
  }
  return call("scan_exposure", { request: { projectId, deep } });
}

const demoProof: RedactionProof = {
  checks: [
    { name: "inspect", passed: true },
    { name: "exposure-scan", passed: true },
    { name: "redacted-occurrences", passed: true },
    { name: "inspect-after-write", passed: true },
    { name: "guide-roundtrip", passed: true },
  ],
  passed: 5,
  total: 5,
  durationMs: 12,
};

export async function runRedactionSelfCheck(): Promise<RedactionProof> {
  if (!isTauriRuntime) {
    return demoProof;
  }
  return call("run_redaction_self_check");
}

File: crates/env-core/src/exposure.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ExposureSeverity {
    Certain,
    Likely,
    Possible,
}

impl ExposureSeverity {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Certain => "certain",
            Self::Likely => "likely",
            Self::Possible => "possible",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ExposureDisposition {
    Exposed,
    Allowed,
    Managed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ExposureKind {
    EnvFile,
    DevVars,
    CredentialFile,
    Npmrc,
    Pypirc,
    Netrc,
    SshPrivateKey,
    AwsCredentials,
    McpConfig,
    GoogleServices,
    ShellHistory,
    AgentTranscript,
}

impl ExposureKind {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::EnvFile => "env-file",
            Self::DevVars => "dev-vars",
            Self::CredentialFile => "credential-file",
            Self::Npmrc => "npmrc",
            Self::Pypirc => "pypirc",
            Self::Netrc => "netrc",
            Self::SshPrivateKey => "ssh-private-key",
            Self::AwsCredentials => "aws-credentials",
            Self::McpConfig => "mcp-config",
            Self::GoogleServices => "google-services",
            Self::ShellHistory => "shell-history",
            Self::AgentTranscript => "agent-transcript",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExposureFinding {
    pub path: String,
    pub kind: ExposureKind,
    pub severity: ExposureSeverity,
    pub disposition: ExposureDisposition,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

impl ExposureFinding {
    pub fn id(&self) -> String {
        format!("{}:{}", self.kind.as_str(), self.path)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ExposureState {
    Scanned,
    Unavailable,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExposureCounts {
    pub certain: usize,
    pub likely: usize,
    pub possible: usize,
    pub allowed: usize,
    pub managed: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExposureProjection {
    pub state: ExposureState,
    pub deep: bool,
    pub counts: ExposureCounts,
    pub findings: Vec<ExposureFinding>,
}

impl ExposureProjection {
    pub fn unavailable(deep: bool) -> Self {
        Self {
            state: ExposureState::Unavailable,
            deep,
            counts: ExposureCounts::default(),
            findings: Vec::new(),
        }
    }

    pub fn from_findings(deep: bool, findings: Vec<ExposureFinding>) -> Self {
        let mut counts = ExposureCounts::default();
        for finding in &findings {
            match finding.disposition {
                ExposureDisposition::Allowed => counts.allowed += 1,
                ExposureDisposition::Managed => counts.managed += 1,
                ExposureDisposition::Exposed => match finding.severity {
                    ExposureSeverity::Certain => counts.certain += 1,
                    ExposureSeverity::Likely => counts.likely += 1,
                    ExposureSeverity::Possible => counts.possible += 1,
                },
            }
        }
        Self {
            state: ExposureState::Scanned,
            deep,
            counts,
            findings,
        }
    }
}
